### PR TITLE
Improve Dockerfile.

### PR DESCRIPTION
**Pull request by schmmd**

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,13 @@
+.dockerignore
+*.pyc
```
Comment (schmmd) [inline .dockerignore]: @jkinkead none of these seem to exclude `pyc` files from my docker builds.  The other ignores work fine.

Comment (jkinkead) [inline .dockerignore]: The files get copied to the docker build environment? They're not just getting recreated?

That seems surprising that the other wildcard patterns would work. Does a single-star not expand to characters, maybe?

Comment (schmmd) [inline .dockerignore]: DOH!  You nailed it with "recreated".  Thank you @jkinkead...

**Comment by schmmd** [comment]
@jkinkead mind taking another look?  Any general suggestions?

You mentioned something about cleaning up the apt-get cache, but I didn't see that it was something people do regularly (or what I should do exactly).

**Comment by jkinkead** [comment]
`rm -rf /var/lib/apt/lists/*`, like here:
https://github.com/docker-library/python/blob/c9954b06c8b178d7888bc1626bed5a14e43a9203/3.6/stretch/Dockerfile#L31

File: .dockerignore [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.dockerignore
+**.pyc
+**/__pycache__
+.gitignore
+.git
+./allennlp/data/fields/__pycache__
```
Comment (jkinkead) [inline .dockerignore]: Does the above `**/__pycache__` not catch this?

File: Dockerfile.cpu [hunk 3]
```diff
@@ -1,6 +1,14 @@
 FROM ubuntu:16.04
 
+ENV LC_ALL=C.UTF-8
+ENV LANG=C.UTF-8
+ENV PATH /opt/conda/bin:$PATH
+ENV PYTHONHASHSEED 2157
+
+WORKDIR /stage
 EXPOSE 8000
+ENTRYPOINT ["/bin/sh", "-c"]
```
Comment (jkinkead) [inline Dockerfile.cpu]: I would use the default entrypoint unless you're using this for shell scripting.

File: Dockerfile.cpu [hunk 4]
```diff
@@ -26,35 +30,22 @@ RUN echo 'export PATH=/opt/conda/bin:$PATH' > /etc/profile.d/conda.sh && \
...
-WORKDIR /stage
 
+# Create a conda environment so we have a reproducible version of Python
 RUN conda create -n allennlp --yes python=3.5
-ENV PYTHONHASHSEED 2157
-ENV LC_ALL=C.UTF-8
-ENV LANG=C.UTF-8
+
+# Make sure we enter the allennlp Conda environment automatically
+RUN echo "source activate allennlp" >> ~/.bashrc
```
Comment (jkinkead) [inline Dockerfile.cpu]: Chain with the previous RUN (install conda)?

File: Dockerfile.cpu [hunk 5]
```diff
@@ -26,35 +30,22 @@ RUN echo 'export PATH=/opt/conda/bin:$PATH' > /etc/profile.d/conda.sh && \
     wget --quiet https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda.sh && \
     /bin/bash ~/miniconda.sh -b -p /opt/conda && \
     rm ~/miniconda.sh
-ENV PATH /opt/conda/bin:$PATH
-
-# Setup AllenNLP
-WORKDIR /stage
 
+# Create a conda environment so we have a reproducible version of Python
 RUN conda create -n allennlp --yes python=3.5
```
Comment (jkinkead) [inline Dockerfile.cpu]: Chain with previous RUN (install conda)?

File: Dockerfile.cpu [hunk 6]
```diff
@@ -26,35 +30,22 @@ RUN echo 'export PATH=/opt/conda/bin:$PATH' > /etc/profile.d/conda.sh && \
     wget --quiet https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda.sh && \
     /bin/bash ~/miniconda.sh -b -p /opt/conda && \
     rm ~/miniconda.sh
-ENV PATH /opt/conda/bin:$PATH
-
-# Setup AllenNLP
-WORKDIR /stage
 
+# Create a conda environment so we have a reproducible version of Python
```
Comment (jkinkead) [inline Dockerfile.cpu]: Why are you installing conda at all? Docker is a self-contained environment already - you can just `pip install` stuff locally without needing the virtual env while using an appropriate python base image.

File: kubernetes-webdemo.yaml [hunk 7]
```diff
@@ -14,7 +14,7 @@ spec:
     spec:
       containers:
         - name: allennlp-webdemo
-          image: "896129387501.dkr.ecr.us-west-2.amazonaws.com/allennlp/allennlp-cpu:98d99ea096d0b4944296adac5fe9fc96f9383c14"
+          image: "allennlp/allennlp-cpu:latest"
```
Comment (jkinkead) [inline kubernetes-webdemo.yaml]: Update the image pull policy to "Always" if you're keeping this line.

**Comment by schmmd** [comment]
Thanks @jkinkead.  We've already questioned using conda inside Docker, but we're not doing the work to move off it yet.  How would you pin you version of `python`?

**Comment by jkinkead** [comment]
`FROM python:3.5.4`.

**Comment by jkinkead** [comment]
I'm pretty sure all you need to do is change your `FROM` statement & delete all of the conda refs. You only install things with `pip`, which is in the base Python image.

File: .dockerignore (hunk 2 above)
Comment (schmmd) [inline .dockerignore]: I think they both work.

**Comment by schmmd** [comment]
@jkinkead there's a catch in that we have a parallel [Dockerfile.gpu](https://github.com/allenai/allennlp/blob/master/Dockerfile.gpu) with a different base (nvidia/cuda:8.0-cudnn5-devel).

File: .dockerignore (hunk 2 above)
Comment (joelgrus) [inline .dockerignore]: I also question why we would need a separate line for this particular __pycache__

Comment (schmmd) [inline .dockerignore]: ah oops!  I misunderstood @jkinkead 's comment.  This is just a debugging line I was using earlier.  Your are both absolutely right--it should not be here.

**Pull request merged**